rfc/system/1855-type-metadata.md

```diff
@@ -0,0 +1,130 @@
+# [RR] Typing Metadata
+
+**Authors:**
+
+- @maximsmol
+- @kennyworkman
+
+## 1 Executive Summary
+
+A small patch to introduce flytekit support for python-native type metadata.
+
+## 2 Motivation
+
+Users of flyte wish to have access to arbitrary parameter metadata specified in
+typing annotations.
+
+The `LiteralType` definition in flyteidl already has a `metadata` field in its
+message definition
+(https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/core/types.proto#L89),
+suggesting that support for metadata annotations for flyte types was in the
+works.
+
+We wish to add python-native support for this field by allowing flytekit to
+recognize and parse the use of
+[`typing.Annotation`](https://docs.python.org/3/library/typing.html#typing.Annotated).
+
+This could be to change the presentation of a form input ingesting typed values,
+surface descriptions to users or store indexes to control the order of
+parameters with custom frontends.
+
+This feature will not send metadata to the container at runtime when a task is
+invoked.  If the container has access to the original code, it will be able to
+access/load metadata from that. This is particularly important to note for
+TaskTemplate-based flytekit plugins.
+
+An example of how we would use this annotation to tag a parameter with
+presentation information:
+
+
+```
+amplicon_min_alignment_score: Optional[
+        List[
+            Annotated[
+                int,
+                FlyteMetadata(
+                    {
+                        "appearance": {
+                            "multiselect": {
+                                "options": [50, 60, 70, 80, 90, 100],
+                                "allow_custom": True,
+                            }
+                        }
+                    }
+                ),
+            ]
+        ]
+    ],
+```
+
+## 3 Proposed Implementation
+
+Expansion of the `to_literal_type` method in the core [`TypeEngine`
+class](https://github.com/flyteorg/flytekit/blob/master/flytekit/core/type_engine.py#L341).
+
+```
+    @classmethod
+    def to_literal_type(cls, python_type: Type) -> LiteralType:
+        """
+        Converts a python type into a flyte specific ``LiteralType``
+        """
+        transformer = cls.get_transformer(python_type)
+        res = transformer.get_literal_type(python_type)
+        meta = None
+        if hasattr(python_type, "__metadata__"):
+            for x in python_type.__metadata__:
+                if not isinstance(x, FlyteMetadata):
+                    continue
+                if x.data.get("__consumed", False):
+                    continue
+                meta = x.data
+                x.data["__consumed"] = True
+
+        if meta is not None:
```

Comment:
One problem with this approach is that `metadata` is currently used to store the json schema for struct types. I think we should either change that behavior (which will be breaking) or introduce a new field for storing documentation more uniformly in the LiteralType?

cc @EngHabu

My preference is to introduce a new substructure called Documentation to store

Reply:
If you want to reserve metadata for internal flyte usage, Annotation makes more semantic sense to me ( as a field that can be overloaded by clients)